metaseq/tasks/streaming_language_modeling.py

```diff
@@ -211,24 +212,41 @@ def _alpha_sampling(self, datasets, corpora, epoch=1):
             [len(d) for d in datasets],
             dtype=float,
         )
+        sample_probs = self._get_sample_prob(dataset_lengths)
         logger.info(f"loaded total {dataset_lengths.sum()} blocks for all corpora")
-        if self.args.pretrain_data_sampling_prob>0:
-            assert self.args.pretrain_data_sampling_prob < 1.0
-            def prefix_match_index(l):
-                prefix = "pretrain__"
-                for i, s in enumerate(l):
-                  if prefix in s:
-                      return i
-                return -1
-            pretrain_corpus_index = prefix_match_index(corpora)
-            assert pretrain_corpus_index != -1
-            tmp_dataset_lengths = np.copy(dataset_lengths)
-            tmp_dataset_lengths[pretrain_corpus_index] = 0
-            sample_probs = self._get_sample_prob(tmp_dataset_lengths)
-            sample_probs = sample_probs * (1 - self.args.pretrain_data_sampling_prob)
-            sample_probs[pretrain_corpus_index] = self.args.pretrain_data_sampling_prob
-        else:
-            sample_probs = self._get_sample_prob(dataset_lengths)
+        per_benchmark_prob = {}
+        for i, s in enumerate(corpora):
+            benchmark = s.split('__')[0]
+            if benchmark not in per_benchmark_prob:
+                per_benchmark_prob[benchmark] = 0.0
+            per_benchmark_prob[benchmark] += sample_probs[i]
+        logger.info('Initial data proportions:')
+        logger.info(json.dumps(per_benchmark_prob))
+
+        data_sampling_prob = json.loads(self.args.data_sampling_prob)
+
+        # If proportions are not specified for all the benchmarks
+        # scale up/down the remaining proportions uniformly
+        total_benchmarks = len(per_benchmark_prob)
+        total_assigned_prob = sum(data_sampling_prob.values())
+        total_assigned_benchmarks = len(data_sampling_prob)
+        for benchmark in per_benchmark_prob:
+            if benchmark not in data_sampling_prob:
+                data_sampling_prob[benchmark] = per_benchmark_prob[benchmark] * (1.0 - total_assigned_prob)
+
+        for i, s in enumerate(corpora):
+            benchmark = s.split('__')[0]
+            sample_probs[i] = sample_probs[i] * data_sampling_prob[benchmark] / per_benchmark_prob[benchmark]
+
+        new_per_benchmark_prob = {}
+        for i, s in enumerate(corpora):
+            benchmark = s.split('__')[0]
+            if benchmark not in new_per_benchmark_prob:
+                new_per_benchmark_prob[benchmark] = 0.0
+            new_per_benchmark_prob[benchmark] += sample_probs[i]
```

Comment:
Can we add some more assertions like new_per_benchmark_prob matches the intended proportions given in data_sampling_prob?